web/ts/edit-course.ts

```diff
@@ -620,44 +727,93 @@ export function createLectureForm(args: { s: [] }) {
                         this.error = true;
                     });
             }
         },
-        uploadVod() {
-            const xhr = new XMLHttpRequest();
-            const vodUploadFormData = new FormData();
-            vodUploadFormData.append("file", this.formData.file[0]);
-            xhr.onloadend = () => {
-                if (xhr.status === 200) {
-                    window.location.reload();
-                } else {
-                    this.error = true;
-                }
-            };
-            xhr.upload.onprogress = (e: ProgressEvent) => {
-                if (!e.lengthComputable) {
-                    return;
-                }
-                window.dispatchEvent(
-                    new CustomEvent("voduploadprogress", { detail: Math.floor(100 * (e.loaded / e.total)) }),
-                );
+
+        async uploadVod(): Promise<number> {
+            const uploadPres = this.formData.presFile[0] != null;
+            const uploadCam = this.formData.camFile[0] != null;
+
+            let uploadProgress = {
+                COMB: 0,
+                PRES: uploadPres ? 0 : null,
+                CAM: uploadCam ? 0 : null,
             };
-            xhr.open(
-                "POST",
-                `/api/course/${this.courseID}/uploadVOD?start=${this.formData.start}&title=${this.formData.title}`,
+
+            window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
+
+            const { streamID } = JSON.parse(
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVOD?start=${this.formData.start}&title=${this.formData.title}`,
+                    this.formData.combFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, COMB: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
+                    },
+                ),
             );
-            xhr.send(vodUploadFormData);
+
+            if (uploadPres)
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVODMedia?streamID=${streamID}&videoType=PRES`,
+                    this.formData.presFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, PRES: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
+                    },
+                );
+
+            if (uploadCam)
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVODMedia?streamID=${streamID}&videoType=CAM`,
+                    this.formData.camFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, CAM: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
```

Comment:
Can you refactor this to its own method to avoid duplication?

Reply:
not sure how to do it in a clean way in a function, because I have the `uploadProgress` state that is caried through the calls, but I abstracted in a way that I have an array of mediaUploads now and iterate through, so the `uploadFilePost`-Part is there only once. 👍